python/dgl/cuda/nccl.py

```diff
@@ -0,0 +1,143 @@
+"""API creating NCCL communicators."""
+
+from .. import backend as F
+from .._ffi.function import _init_api
+
+_COMM_MODES_MAP = {
+    'remainder': 0
+}
+
+class UniqueId(object):
+    """ Class for allowing python code to create and communicate NCCL Unique
+        IDs, needed for creating communicators.
+    """
+    def __init__(self, id_str=None):
+        """ Create an object reference the current NCCL unique id.
+        """
+        if id_str:
+            if isinstance(id_str, bytes):
+                id_str = id_str.decode('utf-8')
+            self._handle = _CAPI_DGLNCCLUniqueIdFromString(id_str)
+        else:
+            self._handle = _CAPI_DGLNCCLGetUniqueId()
+
+    def get(self):
+        """ Get the C-handle for this object.
+        """
+        return self._handle
+
+    def __str__(self):
+        return _CAPI_DGLNCCLUniqueIdToString(self._handle)
+
+    def __repr__(self):
+        return "UniqueId[{}]".format(str(self))
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+
+class Communicator(object):
+    """ High-level wrapper for NCCL communication.
+    """
+    def __init__(self, size, rank, unique_id):
+        """ Create a new NCCL communicator.
+
+            Parameters
+            ----------
+            size : int
+                The number of processes in the communicator.
+            rank : int
+                The rank of the current process in the communicator.
+            unique_id : NCCLUniqueId
+                The unique id of the root process (rank=0).
+        """
+        assert rank < size
+        self._handle = _CAPI_DGLNCCLCreateComm(size, rank, unique_id.get())
+        self._rank = rank
+        self._size = size
+
+    def sparse_all_to_all_push(self, idx, value, mode):
+        """ Perform an all-to-all-v operation, where by all processors send out
+            a set of indices and corresponding values.
+
+            Parameters
+            ----------
+            idx : IdArray
+                The 1D set of indices to send to other processors.
+            value : NDArray
+                The multi-dimension set of values to send to other processors.
+                The 0th dimension must match that of `idx`.
+            mode : int
+                The method for assigning indices to processors.
+
+            Returns
+            -------
+            IdArray
+                The set of recieved indices.
+            NDArray
+                The set of recieved values.
+
+        """
+        mode_id = _COMM_MODES_MAP[mode]
+
+        out_idx, out_value = _CAPI_DGLNCCLSparseAllToAllPush(
+            self.get(), F.zerocopy_to_dgl_ndarray(idx),
+            F.zerocopy_to_dgl_ndarray(value),
+            mode_id)
+        return (F.zerocopy_from_dgl_ndarray(out_idx),
+                F.zerocopy_from_dgl_ndarray(out_value))
+
+    def sparse_all_to_all_pull(self, req_idx, value, mode):
```

Comment:
The API will be changed if we introduce distributed training.
Can we break this API into multiple APIs including 1) get req_idx split according to mode; 2) play alltoallv to synchronize tensors; 
Then when we support distributed all_to_all_pull and all_to_all_push, we only need to provide the distributed version of req_idx split (according to mode and the partition book).

Reply:
We could just add an assignment vector as your saying so the method would look like:
```
def sparse_all_to_all_pull(self, proc_id, req_idx, value):
```
Where `proc_id` and `req_idx` have the same dimensions.

Another option would be, we could change the `mode` parameter into something like a `Partitioning` object that would have a C++ method the communicator could call:
```
def sparse_all_to_all_pull(self, req_idx, value, partitioning):
```
```
template<typename IdType>
void Partitioning::Map(const IdType * idx, int64_t num, IdType * proc, cudaStream_t stream);
```

That way we could have tighter control of stream and memory management, as well as avoid the overheads associated with going back and forth with python.